metrics-core/src/main/java/com/codahale/metrics/ExponentiallyDecayingReservoir.java

```diff
@@ -92,26 +89,11 @@ public void update(long value) {
      */
     public void update(long value, long timestamp) {
         rescaleIfNeeded();
-        lockForRegularUsage();
-        try {
-            final double itemWeight = weight(timestamp - startTime);
-            final WeightedSample sample = new WeightedSample(value, itemWeight);
-            final double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
-
-            final long newCount = count.incrementAndGet();
-            if (newCount <= size || values.isEmpty()) {
-                values.put(priority, sample);
-            } else {
-                Double first = values.firstKey();
-                if (first < priority && values.putIfAbsent(priority, sample) == null) {
-                    // ensure we always remove an item
-                    while (values.remove(first) == null) {
-                        first = values.firstKey();
-                    }
-                }
-            }
-        } finally {
-            unlockForRegularUsage();
+        final State localState = this.writerState;
+        localState.update(value, timestamp);
+        final State newLocalState = this.writerState;
+        if (localState != newLocalState) {
```

Comment:
I am not sure this is needed but the idea is that it might help us avoid losing events that happen at the same time as a rescale operation triggered by another thread